### PR TITLE
Lpabon/6138 circular schema protection

### PR DESCRIPTION
**Pull request by TurkeyFried**
This PR adds protection for infinite loops when running the FHIR to HL7 translator.

Test Steps:
1. Check out the branch for this PR.
2. Run the following command to convert a FHIR bundle to HL7.  
`./prime fhirdata --input-file "src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir" -s metadata/hl7_mapping/ORU_R01/ORU_R01-base.yml --output-format=FHIR`
3. Verify an HL7 message is generated from the conversion.
4. To force an infinite loop, go to `order-observation.yml` and replace Line 16 to `schema: order-observation`
5. Run the translator again. This time, it will throw an `HL7ConversionException` with message `Circular reference detected in schema`

## Changes
- Add checks on the recursive function that handles schema inheritance.
- Fixes typo on an existing config file.

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #6138

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/metadata/hl7_mapping/ORU_R01/base/entity-identifier.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: ORU-R01-order
+name: ORU-R01-entity-identifier
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/entity-identifier.yml:1]: Thank you!

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchema.kt [hunk 2]
```diff
@@ -22,6 +22,8 @@ data class ConfigSchema(
     var constants: SortedMap<String, String> = sortedMapOf(),
     var extends: String? = null
 ) {
+    var ancestry: List<String> = listOf()
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchema.kt:25]: Ancestry.com!!! ;D Good name!

Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchema.kt:25]: Needs kdoc here.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt [hunk 3]
```diff
@@ -70,10 +75,13 @@ object ConfigSchemaReader : Logging {
             throw SchemaException(msg, e)
         }
 
+        if (ancestry.contains(rawSchema.name)) throw HL7ConversionException("Circular reference detected in schema")
+        rawSchema.ancestry = ancestry + rawSchema.name!!
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:79]: I did not know you could do that!

Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:79]: I think you can just move this down to line 84 without using rawSchema.ancestry.  I do not think there is a need for rawSchema.ancestry at that point.

**Review by carlosfelix2 — APPROVED** [review]
Tested locally

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt (hunk 3 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:79]: Gave that a shot, but it doesn't work without having the object reference unless plenty more is done

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchema.kt (hunk 2 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchema.kt:28]: done

**Review by clediggins-usds — APPROVED** [review]

**Pull request merged**